juju/juju
# [JUJU-3749] Support bases in bundles

Use bases throughout bundle deploy code, parsing immidiately from series
if necessary

Unfortunately, the least bad option was to keep bases as strings in
Change structs. This means the marshalled nicely for Args as a string, and
omitempty is respected

Also, since GUIArgs output ordering is magic, we cannot replace series
with base. GUIArgs and the associated GetChanges facade endpoint should
be deprecated in favour of GetChangesMapArgs in Juju 4

## Checklist


## QA steps

Ensure unit tests and test_deploy_bundles integration tests pass

```
$ juju add-model m1
$ cat bundle1.yaml 
name: tiny
applications:
  ubuntu-plus-one:
    charm: ./testcharms/charms/ubuntu-plus
    scale: 1
    base: "ubuntu@20.04"
  ubuntu-plus-two:
    charm: ./testcharms/charms/ubuntu-plus
    scale: 1
    base: "ubuntu@22.04"
$ juju deploy ./bundle1.yaml
Executing changes:
- upload charm /home/jack/juju/juju-core/testcharms/charms/ubuntu-plus for base ubuntu@20.04/stable with architecture=amd64
- deploy application ubuntu-plus-one on ubuntu@20.04/stable
- upload charm /home/jack/juju/juju-core/testcharms/charms/ubuntu-plus for base ubuntu@22.04/stable with architecture=amd64
- deploy application ubuntu-plus-two on ubuntu@22.04/stable
- add unit ubuntu-plus-one/0 to new machine 0
- add unit ubuntu-plus-two/0 to new machine 1
Deploy of bundle completed.

(wait)
$ juju status
Model  Controller  Cloud/Region         Version      SLA          Timestamp
m      lxd         localhost/localhost  3.3-beta1.1  unsupported  12:02:21+01:00

App              Version  Status  Scale  Charm        Channel  Rev  Exposed  Message
ubuntu-plus-one           active      1  ubuntu-plus             0  no       Hello from update-status, load: 0.37, 0.79, 1.52
ubuntu-plus-two           active      1  ubuntu-plus             0  no       Hello from update-status, load: 0.70, 0.83, 1.50

Unit                Workload  Agent  Machine  Public address  Ports  Message
ubuntu-plus-one/0*  active    idle   0        10.219.211.227         Hello from update-status, load: 0.37, 0.79, 1.52
ubuntu-plus-two/0*  active    idle   1        10.219.211.32          Hello from update-status, load: 0.70, 0.83, 1.50

Machine  State    Address         Inst id        Base          AZ  Message
0        started  10.219.211.227  juju-738e43-0  ubuntu@20.04      Running
1        started  10.219.211.32   juju-738e43-1  ubuntu@22.04      Running
```

```
$ juju add-model m2
$ cat bundle2.yaml
name: tiny
applications:
  ubuntu-plus-one:
    charm: ./testcharms/charms/ubuntu-plus
    scale: 1
    to:
      - 0
  ubuntu-plus-two:
    charm: ./testcharms/charms/ubuntu-plus
    scale: 1
    to:
      - 1
machines:
  0:
    base: ubuntu@22.04
  1:
    base: ubuntu@20.04
$ juju deploy ./bundle2.yaml
(wait)
$ juju status 
Model  Controller  Cloud/Region         Version      SLA          Timestamp
m2     lxd         localhost/localhost  3.3-beta1.1  unsupported  12:02:21+01:00

App              Version  Status  Scale  Charm        Channel  Rev  Exposed  Message
ubuntu-plus-one           active      1  ubuntu-plus             0  no       Hello from update-status, load: 0.37, 0.79, 1.52
ubuntu-plus-two           active      1  ubuntu-plus             0  no       Hello from update-status, load: 0.70, 0.83, 1.50

Unit                Workload  Agent  Machine  Public address  Ports  Message
ubuntu-plus-one/0*  active    idle   0        10.219.211.227         Hello from update-status, load: 0.37, 0.79, 1.52
ubuntu-plus-two/0*  active    idle   1        10.219.211.32          Hello from update-status, load: 0.70, 0.83, 1.50

Machine  State    Address         Inst id        Base          AZ  Message
0        started  10.219.211.141  juju-c5c5b8-0  ubuntu@22.04      Running
1        started  10.219.211.102  juju-c5c5b8-1  ubuntu@20.04      Running
```

```
$ juju add-model m3 
$ cat bundle3.yaml
name: tiny
default-base: ubuntu@20.04
applications:
  ubuntu-plus-one:
    charm: ./testcharms/charms/ubuntu-plus
    scale: 1
  ubuntu-plus-two:
    charm: ./testcharms/charms/ubuntu-plus
    scale: 1
    base: ubuntu@22.04
$ juju deploy ./bundle3.yaml
Executing changes:
- upload charm /home/jack/juju/juju-core/testcharms/charms/ubuntu-plus for base ubuntu@20.04/stable with architecture=amd64
- deploy application ubuntu-plus-one on ubuntu@20.04/stable
- upload charm /home/jack/juju/juju-core/testcharms/charms/ubuntu-plus for base ubuntu@22.04/stable with architecture=amd64
- deploy application ubuntu-plus-two on ubuntu@22.04/stable
- add unit ubuntu-plus-one/0 to new machine 0
- add unit ubuntu-plus-two/0 to new machine 1
Deploy of bundle completed.

(wait)
$ juju status
Model  Controller  Cloud/Region         Version      SLA          Timestamp
m3     lxd         localhost/localhost  3.3-beta1.1  unsupported  12:02:21+01:00

App              Version  Status  Scale  Charm        Channel  Rev  Exposed  Message
ubuntu-plus-one           active      1  ubuntu-plus             0  no       Hello from update-status, load: 0.37, 0.79, 1.52
ubuntu-plus-two           active      1  ubuntu-plus             0  no       Hello from update-status, load: 0.70, 0.83, 1.50

Unit                Workload  Agent  Machine  Public address  Ports  Message
ubuntu-plus-one/0*  active    idle   0        10.219.211.227         Hello from update-status, load: 0.37, 0.79, 1.52
ubuntu-plus-two/0*  active    idle   1        10.219.211.32          Hello from update-status, load: 0.70, 0.83, 1.50

Machine  State    Address         Inst id        Base          AZ  Message
0        started  10.219.211.227  juju-738e43-0  ubuntu@20.04      Running
1        started  10.219.211.32   juju-738e43-1  ubuntu@22.04      Running
```
## Documentation changes

Base needs to be added to bundle documentation once 3.3 is released
